Avoid cumulative drift in PeriodicThread

PeterBowman opened the pull request:
Fixes https://github.com/robotology/yarp/issues/2488. Instead of a relative value obtained from the time elapsed on each step, which was prone to cause error accumulation over time, now I am using an absolute initial reference that helps getting the exact delay to prevent just that. This reference needs to be reset whenever the period is changed. In short, the new computed delay boils down to:

```cxx
delay = absolute_reference + period * iterations_since_absolute_reference - now;
```

In addition, I have added new checks to PeriodicThreadTest, which I'm not sure why weren't there, for testing the timer accuracy.

Please review. I have tested this on the code snippet attached in https://github.com/robotology/yarp/issues/2488 (using a system clock), and also on a similar example app with `yarp::os::Timer` instead.

drdanz commented:
Sorry if it took so long but I was quite busy, and I want to examine this properly, because I consider this a very delicate matter...

At a first impression I have some concerns

1. The `suspend` and `resume` should probably "adapt" the schedule as well, otherwise after a long suspension, the thread will just start running without even considering the period.

2. The same is valid for runs which for some reason take more time than the period. Also in this case the thread will start running without considering the period. I'm not sure how to handle this one.
 
3. The counter is a potential issue (even if unlikely, it might reach its max and reset to zero), forces you to do a multiplication at every run, and makes the size of the class unnecessarily bigger. What do you think about just storing the time when the next run is supposed to be executed, and forgetting about the reftime? Maybe also issue a warning when the time for the next execution is already passed (perhaps in that case skip the execution? I don't know...). But this would be an issue for PeriodicThreads with period=0


My opinion is that just changing current behaviour is rather dangerous, this change might cause unexpected behaviours in several places... I suggest to add an option to change the behaviour instead, leaving current behaviour as default for now, and perhaps in a second moment, when this behaviour is tested enough, we can eventually change the default to this beaviour.

CC @pattacini, @randaz81, @vtikha, @Nicogene, @lornat75

drdanz commented:
CC @traversaro

pattacini commented:
>The same is valid for runs which for some reason take more time than the period. Also in this case the thread will start running without considering the period. I'm not sure how to handle this one.

If the period gets somehow exceeded, then the thread should start over immediately.

>My opinion is that just changing current behaviour is rather dangerous, this change might cause unexpected behaviours in several places... I suggest to add an option to change the behaviour instead, leaving current behaviour as default for now, and perhaps in a second moment, when this behaviour is tested enough, we can eventually change the default to this behavior.

Ok for the smooth transition 👍🏻

drdanz commented:
> If the period gets somehow exceeded, then the thread should start over immediately afterward.

@pattacini I agree for the following run, but what about the one after that, and the ones after?

Just to give a better explanation about my concern, let's assume that there are 2 concurrent threads accessing the same resource, locking a mutex for the whole run, with a period that is longer than the average run, and therefore it is not causing issues right now.

If for some reason one of the two threads is delayed for some reason, it will start running without stopping, and therefore starving the other thread that will not be able to lock the mutex until the number of executions is back to the expected (or until for a very lucy case the scheduler will reschedule between the unlock and lock). Meanwhile the other one will accumulate delay and as soon as he gets the mutex it will start starving the other. So the threads will start alternating with an unexpected pattern, and this might lead to weird behaviours, data missing, etc...

pattacini commented:
Two periodic threads that need to share a common resource without having enough time margin to complete the round is a clear road to ruin, meaning that starvation is quite an expected happening.

I might have a too real-time-minded approach, but I tend to consider that time budget allocation is a crucial concern.

Anyway, I believe that for consistency we could adopt the official policy of `std::thread`, whatever it is 😃

drdanz commented:
> Two periodic threads that need to share a common resource without having enough time margin to complete the round is a clear road to ruin, meaning that starvation is quite an expected happening.

No sorry, the point is if actually the opposite... They have both enough time to complete the job, but they will start starving each other if one is delayed.

pattacini commented:
So, it's not that the `run()` takes longer than the period but rather it's the start of the `run()` that can get delayed. This may clearly happen because of the scheduler. However, for control purposes, that is right when it's paramount to comply with a stable clock, the developer should take into account that the max duration of the `run()` shall be smaller (with a margin) than the assigned period plus the variability of the delay. That's the reason why going faster than 10 ms is not advisable, sort of (of course depending on the number of operations that need to be done). In this kind of context, using shared resources is also critical anyway, and thus needs to be carefully judged.

It very much depends on the context, eventually: when we do control or when we don't care about tight timing.

Do we know what policy `std::thread` apply?

traversaro commented:
> Do we know what policy `std::thread` apply?

I don't think `std::thread` has any support for periodic thread or similar.

pattacini commented:
>I don't think std::thread has any support for periodic thread or similar.

That's true!

Found out a quite nice although old [resource][1] on this (with illustrations).

Citing it:
>Again, the specific application dictates whether the timing inaccuracy is catastrophic or merely a nuisance to be ignored.

[1]: https://bulldozer00.blog/2013/12/27/periodic-processing-with-standard-c11-facilities/

lornat75 commented:
I agree with the smooth transition.

About this:

> If the period gets somehow exceeded, then the thread should start over immediately afterward.

I admit I haven't read the patch, but to avoid starvation problems (or actually delegating them to the scheduler policies) it should be sufficient to yield the cpu (in windows sleep(0) was actually do this, at the time I was testing this stuff, so quite some time ago), can't this be done/keep this behavior in this specific case?

pattacini commented:
>I admit I haven't read the patch, but to avoid starvation problems (or actually delegating their arbitration to the scheduler policies) it should be sufficient to yield the cpu (in windows sleep(0) was actually do this, at the time I was testing this stuff, so quite some time ago), can't this be done/keep this behavior in this specific case?

When I said `immediately afterward` I meant from the thread standpoint, that is without requiring the addition of any further delay on purpose within the code handling the periodicity.

A call to `yarp::osTime::yield()` should be doing exactly this, actually: it releases the quantum without touching the periodicity so that the thread will be run again as soon as the schedule will permit to do so.

PeterBowman commented:
> 1. The `suspend` and `resume` should probably "adapt" the schedule as well, otherwise after a long suspension, the thread will just start running without even considering the period.

The `run` method is entered regardless of the thread being suspended or not, therefore the new period will be adapted (and the new delay applied at the end) anyway. I'm not sure if I understood your comment correctly, though. Is it about calling `setPeriod()` while suspended?

> 2. The same is valid for runs which for some reason take more time than the period. Also in this case the thread will start running without considering the period. I'm not sure how to handle this one.

The next step will start immediately after the previous one, and this should also be the case for the current implementation (negative `sleepPeriod`). For the following steps though, it depends on how long the *busy* step was and how much intermediate steps contributed to the accumulated elapsed time. So if the busy step takes like three periods to complete, then steps 2, 3 and 4 will start immediately, and perhaps step 5 too if the combined duration of 2+3+4 exceeds the fixed period.

> 3. The counter is a potential issue (even if unlikely, it might reach its max and reset to zero), forces you to do a multiplication at every run, and makes the size of the class unnecessarily bigger. What do you think about just storing the time when the next run is supposed to be executed, and forgetting about the reftime? Maybe also issue a warning when the time for the next execution is already passed (perhaps in that case skip the execution? I don't know...). But this would be an issue for PeriodicThreads with period=0

I don't think the refTime could be skipped, it is necessary to get an accurate estimation of successive step executions. Calculating the next run on the fly would kind of replicate the issues I experienced with the current implementation (error accumulation). The class isn't much bigger, I'm trading the `elapsed` (forgot to delete this) and `sleepPeriod` members for `refTime`, `countOffset` and the Boolean `scheduleAdapt`. Multiplications are already performed on each step to obtain statistics. I know this will sound silly, but an `unsigned long (long)` counter would take like forever to overflow even with period equal to zero.

> My opinion is that just changing current behaviour is rather dangerous, this change might cause unexpected behaviours in several places... I suggest to add an option to change the behaviour instead, leaving current behaviour as default for now, and perhaps in a second moment, when this behaviour is tested enough, we can eventually change the default to this beaviour.

I'm OK with that if deemed necessary. I'm assuming the breaking change here is the difference in the behavior regarding successive runs that follow a busy step which exceeds the fixed period. This should be pointed out in the docs.

drdanz commented:
Sorry, there is a big misconception here about what `yield`, does.

`yield` does not protect against starvation, and what it does is heavily dependent on the operating system and scheduler in use. `yield` just hints the scheduler that this might be a good moment to schedule a different thread, but the scheduler might decide to reschedule the same thread. Incidentally, on some system this also protects from starving a different thread.

On Linux, with the default scheduler, a thread that unlocks, yields, and locks again will still starve other threads that are waiting on the same resource, because when `yield` is called, the other threads will pass from the blocked queue to the running queue, but the scheduler will run a thread that is already on the running queue, i.e. (in case of 2 threads) the thread that just called `yield`. Then when the time slice for the thread is over, control will go back to the scheduler, that has to choose again, it will see that the other thread is blocked and move it to the blocked queue, and schedule again the first thread... and so on.

This issue was deeply analysed and discussed in the past, we can't rely or give any guarantee in YARP on a behaviour that is system dependent, therefore we have to assume the worst hypothesis here, i.e. yield does not protect against starvation when the whole run is protected by a mutex, and we don't want to to have workarounds that are complicated to maintain, dependent on the system, might make the performances worse, and that encourage bad programming styles (i.e. locking the whole `run`).

drdanz commented:
> > 1. The `suspend` and `resume` should probably "adapt" the schedule as well, otherwise after a long suspension, the thread will just start running without even considering the period.
> 
> The `run` method is entered regardless of the thread being suspended or not, therefore the new period will be adapted (and the new delay applied at the end) anyway.

No, you are right, I did not realized that the method is called anyway...

drdanz commented:
> I don't think the refTime could be skipped, it is necessary to get an accurate estimation of successive step executions. Calculating the next run on the fly would kind of replicate the issues I experienced with the current implementation (error accumulation).

Sorry, I don't understand why this should cause error accumulation, I believe that in each run you just have to sum the last expected run + the period, in order to get the next expected run, with no error...

pattacini commented:
Thanks @drdanz for the clarification on the gut of `yiedl` 👍🏻
However, I'm afraid that all these details are somehow hiding the original main point that in my view can be recapped as follows:
- For time-critical tasks (like in the control domain), we need to favor the scheduling of the operations on the basis of a stable clock (that prevents drifting) over the risk of starvation for threads that do resource sharing. In my understanding (correct me if I'm wrong), it's not doable to guarantee both the objectives at the same time (stable clock + no risk of starvation). If that's the case, then we should favor the former over the latter, as said. it's thus the responsibility of the developer to design accurately the timeline of the operations and how resources need to be allocated in order to limit or even rule out the possibility of starvation (by design, so to speak).
- For other kinds of periodic operations, starvation may be a serious concern.

Perhaps, it'd be convenient to keep the opportunity to have both the strategies available through a configuration option.
What do you think, guys?

PeterBowman commented:
> > I don't think the refTime could be skipped, it is necessary to get an accurate estimation of successive step executions. Calculating the next run on the fly would kind of replicate the issues I experienced with the current implementation (error accumulation).
> 
> Sorry, I don't understand why this should cause error accumulation, I believe that in each run you just have to sum the last expected run + the period, in order to get the next expected run, with no error...

In the following example, the next iteration instant is obtained on each step by simply adding the fixed delay to the previous value. Using a delay of 5 milliseconds over 50000 iterations (=250 seconds), on a test run the loop ended 6.85 milliseconds too late, which exceeds one full period. With `--delay 0.002 --iterations 100000` (200 s) it was 10 ms (5 periods).

<details>
<summary>main.cpp</summary>

```cxx
#include <yarp/os/Log.h>
#include <yarp/os/Property.h>
#include <yarp/os/SystemClock.h>

using namespace yarp::os;

int main(int argc, char * argv[])
{
    yarp::os::Property opts;
    opts.fromCommand(argc, argv);

    const auto delay = opts.check("delay", yarp::os::Value(0.005)).asFloat64(); // seconds
    const auto iterations = opts.check("iterations", yarp::os::Value(10000)).asInt32();

    const auto start = SystemClock::nowSystem();
    auto nextExecution = start;

    for (auto i = 0; i < iterations; i++)
    {
        nextExecution += delay;
        yInfo("[%d] now: %f", i + 1, SystemClock::nowSystem());
        SystemClock::delaySystem(delay / 2.0); // simulate some task going on here
        SystemClock::delaySystem(nextExecution - SystemClock::nowSystem());
    }

    const auto actual = SystemClock::nowSystem();
    const auto expected = start + iterations * delay;
    yInfo("Expected end: %f, actual end: %f, diff: %f", expected, actual, actual - expected);
    return 0;
}
```
</details>

"Error accumulation" was a wrong choice of words from my side, perhaps it'd be rather "timing accuracy". Since clock stability is paramount, I believe that not using an absolute reference will always result in inaccuracies in one way or another.

drdanz commented:
Oh, I see, the error here is not related to the actual timing, but is purely a floating point arithmetic issue.

<details>
<summary>Slightly modified `main.cpp` that does not use time or delay at all</summary>

```c++
#include <yarp/os/Log.h>
#include <yarp/os/Property.h>

using namespace yarp::os;

int main(int argc, char * argv[])
{
    yarp::os::Property opts;
    opts.fromCommand(argc, argv);

    const auto delay = opts.check("delay", yarp::os::Value(0.005)).asFloat64(); // seconds
    const auto iterations = opts.check("iterations", yarp::os::Value(10000)).asInt32();

    const auto start = 1600000000.0;
    auto nextExecution = start;

    for (auto i = 0; i < iterations; i++)
    {
        nextExecution += delay;
        auto nextAbs = (start + (i + 1) * delay);
        yInfo("[%d] nextExecution: relative %.10f, absolute %.10f, diff = %.10f", i + 1, nextExecution, nextAbs, nextExecution - nextAbs);
    }

    return 0;
}
```
</details>

To be honest I wasn't expecting that much error from doubles, but this makes a lot of sense, `double` is not appropriate to represent time...
Anyway, in this case I agree that we need the absolute reference.

Perhaps in the very unlikely case that the counter reaches its max, it should "adjust" and reset the counter?

Anyway, I'm still in favour of adding a setting to choose between current behaviour, and the new one in order to provide a smooth transition.

PeterBowman commented:
I have abstracted out the way the sleep period is obtained with a pair of new classes, representing a stable clock (new) and the usual one (default). This selection can be configured in the `PeriodicThread`'s constructor, now accepting a third (optional) parameter:

```cxx
yarp::os::PeriodicThread thread(period, useSystemClock, yarp::os::ShouldUseStableClock::Yes /* new */);
```

jgvictores commented:
Very cool! Maybe a more semantic alternative could be found for `yarp::os::ShouldUseStableClock::Yes`? Here a couple of proposals for inspiration. 

- `yarp::os::PeriodicThreadClock::Absolute` vs ``yarp::os::PeriodicThreadClock::Relative`
- `yarp::os::PeriodicThreadClock::AvoidCumulativeDrift::Yes`

pattacini commented:
I personally like `yarp::os::PeriodicThreadClock::Absolute` vs `yarp::os::PeriodicThreadClock::Relative`.

drdanz commented on src/libYARP_os/src/yarp/os/PeriodicThread.cpp:
```diff
@@ -182,29 +243,30 @@ class yarp::os::PeriodicThread::Private : public ThreadImpl
 
     void run() override
     {
+        delayEstimator->reset(nowFunc());
```
I'm not really sure, but should `reset` be protected by `lock()` and `unlock()` as well?

PeterBowman replied:
It should not be a problem since `run()` is called only once after thread start. There are other methods that also lack mutex protection such as `getPeriod()`, `resetStat()` and `suspend()/resume()/isSuspended()`. Should I lock everything, just to be 100% sure?

drdanz reviewed (commented):
Thanks a lot @PeterBowman!
The patch looks good to me now, but I agree with @jgvictores and @pattacini that  `yarp::os::PeriodicThreadClock::Absolute` and `yarp::os::PeriodicThreadClock::Relative` is easier to understand.

drdanz commented:
@PeterBowman it would also be great if you could add some test for that use the new constructor parameter (you can just copy one of the existing)

drdanz commented:
I think it would make sense to add also a second constructor

```c++
    explicit PeriodicThread(double period,
                            ShouldUseStableClock useStableClock);
```

that does not take the `useSystemClock` option... Otherwise if you just want to test the new functionality, you need to add 2 parameters

PeterBowman commented:
Update:
* Now using the `yarp::os::PeriodicThreadClock` enum with values `Relative` and `Absolute`.
* Added constructor adapter: `explicit PeriodicThread(double period, PeriodicThreadClock clockAccuracy)`.
* Unit test for `PeriodicThread`'s precision* has been reworked and split into relative and absolute Catch sections. The former aims to ensure the actual period is greater than the requested one (due to the nature of this mode), the latter checks the accuracy within a given threshold, pretty much like it was being done before. However, note the previous implementation was simply printing info and warning logs, probably because a relative clock would never give repeatable nor accurate results. I have dropped logs and turned them to actual `CHECK()` statements.

(*) I am sticking to the term "accuracy" in docs and variable names per [this reference](https://www.ni.com/es-es/support/documentation/supplemental/06/what-clock-error-means-to-your-measurement-system.html):

> (...) There are three main components of clock error -- accuracy, stability, and jitter. The clock accuracy describes how well the actual frequency matches the specified frequency. (...) The clock stability describes how well the oscillator frequency resists fluctuations. (...) Jitter refers to small variations in the period of the clock from one edge to the next, and each additional hardware component on the measurement device adds jitter. (...)

PeterBowman commented:
Last commit fixes a race condition that occurred whenever `setPeriod()` was called during the `yield()` code section between locks, which caused to obtain the sleep period (second lock) without having updated/adapted the internal counter offset (first lock).

Besides, I noticed someone could stop and resume the same thread, yet the `count` member would not be zeroed. The logic behind the "adapt" schedule has been reviewed to ensure the counter offset is properly set for correct delay estimation.

Unit tests now cover the `setPeriod()` case as well. Marking as ready for review.

drdanz reviewed (approved):
LGTM now, thanks a lot!

drdanz commented:
/rebase

drdanz commented:
Merged, thanks.
